### PR TITLE
GH-23: Added Tizen backend

### PR DESCRIPTION
**Pull request by sung-su**
### Description of Change ###

Added Tizen backend. (https://github.com/xamarin/Essentials/issues/23)

Implemented APIs:

 - [x] Accelerometer
 - [x] AppInfo
 - [x] Barometer
 - [x] _Battery_ _(partially supported: energy saver is not supported)_
 - [x] Browser
 - [x] ~~Clipboard~~ _(not supported)_
 - [x] Compass
 - [x] Connectivity
 - [x] ~~DeviceDisplay~~ _(not supported)_
 - [x] DeviceInfo
 - [x] Email
 - [x] FileSystem
 - [x] Flashlight
 - [x] Geocoding
 - [x] Geolocation
 - [x] Gyroscope
 - [x] Launcher
 - [x] Magnetometer
 - [x] MainThread
 - [x] Map
 - [x] OrientationSensor
 - [x] PhoneDialer
 - [x] Platform
 - [x] Preferences
 - [x] SecureStorage
 - [x] Share
 - [x] Sms
 - [x] TextToSpeech
 - [x] VersionTracking
 - [x] Vibration

screenshot_mobile

### CODE REVIEW

**Comment by mattleibow** [comment]
Build

**Comment by sung-su** [comment]
Jenkins failed by 'test-ios-emu'. I think that failure is not related to this PR. Please try build again.

**Comment by sung-su** [comment]
Jenkins failed with 'test-ios-emu' I think that failure is not related to this PR.

**Comment by mattleibow** [comment]
build

**Comment by Redth** [comment]
@sung-su We're going to be introducing some API changes (sorry!).  Let's wait on this until those come in and have you update the PR once that happens.

**Comment by Redth** [comment]
Given we had a number of API changes, you'll need to rebase this on master now.  Sorry!

**Comment by Redth** [comment]
This is related to #23

**Comment by sung-su** [comment]
Rebased to the latest.
Limitation modules are 'Battery', 'Clipboard', and 'Device Display'.

**Comment by Redth** [comment]
@sung-su what are the limitations?

**Comment by sung-su** [comment]
@Redth 
I mean, not support modules on current Tizen platform.
Those are not implemented.

**Comment by rookiejava** [comment]
To be more specific, some features are not in the current Tizen version (4.0), but some can be supported in the future. (5.0 or higher)

**Comment by Redth** [comment]
Got it.  Looked a bit at the PR and it looks like you're generally _doing the right thing_ and throwing exceptions where you have no way to implement.

I'm thinking we should not be throwing `NotImplementedInReferenceAssemblyException` though as that makes the user think they somehow got the netstandard reference instead of the tizen one, which isn't really the case.  I think we should instead use `PlatformNotSupportedException` in these cases.  Maybe you want to define a `const string` somewhere with a message about "This API is not currently supported on Tizen" or something similar and use that as the message.

**Comment by sung-su** [comment]
Changed exception from 'NotImplementedInReferenceAssemblyException' to 'PlatformNotSupportedException' .

**Comment by mattleibow** [comment]
This is an absolute beauty this PR! I really hope that we can merge this in sometime! The code is clean, The APIs are simple and neat, and it just looks like it is meant to be a part of the main repo!

**Comment by mattleibow** [comment]
@sung-su I just realised that we actually have another exception: `FeatureNotSupportedException`. This means that you might need to change the exception for the power saver feature.

So far, we are using 3 exceptions:

 - `PlatformNotSupportedException` - this means the entire API is not supported, such as `DeviceDisplay` and `Clipboard`
 - `FeatureNotSupportedException` - this means that the specific part of the API doesn't work, such as `Battery` with the energy saver
 - `FeatureNotEnabledException` - this means the user must enable a hardware feature, such as the GPS
 - `PermissionException` - this means the user denied the app permission, such as revoking GPS access

@Redth @jamesmontemagno can you confirm?

**Comment by sung-su** [comment]
Changed the exception.

**Comment by mattleibow** [comment]
@sung-su @jamesmontemagno I was looking at the Android code for the battery saver stuff, and I see some versions also do not support this feature. In those cases we do not throw, but just do nothing:
https://github.com/xamarin/Essentials/blob/master/Xamarin.Essentials/Battery/Battery.android.cs#L12-L48

We need to formally decide what we need to do here. I am reopening an issue because this is not going well: https://github.com/xamarin/Essentials/issues/19

**Comment by mattleibow** [comment]
build

**Comment by sung-su** [comment]
@mattleibow 
Does this PR continue to wait until Issue #19 is resolved?

**Comment by jamesmontemagno** [comment]
We probably need to look into the CI server, but we should be able to try to add in Tizen here.. Will work with the team on monday.

**Comment by mattleibow** [comment]
There seems to be a code error:

> Launcher/Launcher.shared.cs(45,38): error CS1503: Argument 1: cannot convert from 'Xamarin.Essentials.OpenFileRequest' to 'System.Uri'

**Comment by sung-su** [comment]
I just rebased and fixed.

File: Xamarin.Essentials/TextToSpeech/TextToSpeech.tizen.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                    if (voice.Language == settings.Locale.Language)
+                    {
+                        language = voice.Language;
+                        voiceType = voice.VoiceType;
+                    }
+                }
+            }
+            var speed = 0;
+            if (settings?.Pitch.HasValue ?? false)
+                speed = (int)settings.Pitch.Value;
```
Comment (Redth) [inline Xamarin.Essentials/TextToSpeech/TextToSpeech.tizen.cs]: Should there be a conversion of some sort here for Pitch?  We treat pitch as a minimum of `0.0f`, normal is `1.0f` and fastest is `2.0f`.

File: Xamarin.Essentials/Barometer/Barometer.tizen.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+        internal static void PlatformStop()
+        {
+            DefaultSensor.DataUpdated -= DataUpdated;
+            DefaultSensor.Stop();
+        }
+
+        static void DataUpdated(object sender, PressureSensorDataUpdatedEventArgs e)
+        {
+            OnChanged(new BarometerData(e.Pressure));
```
Comment (Redth) [inline Xamarin.Essentials/Barometer/Barometer.tizen.cs]: What is your unit of pressure?  We always return `hectopascals`, want to make sure yours is the same.

File: Xamarin.Essentials/Geolocation/Geolocation.tizen.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            cancellationToken = Utils.TimeoutToken(cancellationToken, request.Timeout);
+            cancellationToken.Register(() =>
+            {
+                service?.Stop();
+                tcs.TrySetResult(false);
+            });
+
+            service.LocationChanged += (s, e) =>
+            {
+                if (e.Location != null)
```
Comment (Redth) [inline Xamarin.Essentials/Geolocation/Geolocation.tizen.cs]: Just checking the following:
 - [ ] Accuracy in meters?
 - [ ] Altitude in meters?
 - [ ] Speed in meters per second?

File: Xamarin.Essentials/Magnetometer/Magnetometer.tizen.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+        internal static void PlatformStop()
+        {
+            DefaultSensor.DataUpdated -= DataUpdated;
+            DefaultSensor.Stop();
+        }
+
+        static void DataUpdated(object sender, MagnetometerDataUpdatedEventArgs e)
+        {
+            OnChanged(new MagnetometerData(e.X, e.Y, e.Z));
```
Comment (Redth) [inline Xamarin.Essentials/Magnetometer/Magnetometer.tizen.cs]: Are these values microteslas (µ)?

File: Xamarin.Essentials/Xamarin.Essentials.csproj [hunk 5]
```diff
@@ -1,7 +1,8 @@
 ﻿<Project Sdk="MSBuild.Sdk.Extras/2.0.29">
   <PropertyGroup>
-    <TargetFrameworks Condition=" '$(OS)' == 'Windows_NT' ">netstandard1.0;netstandard2.0;Xamarin.iOS10;MonoAndroid60;MonoAndroid70;MonoAndroid71;MonoAndroid80;MonoAndroid81;MonoAndroid90;uap10.0.16299</TargetFrameworks>
-    <TargetFrameworks Condition=" '$(OS)' != 'Windows_NT' ">netstandard1.0;netstandard2.0;Xamarin.iOS10;MonoAndroid60;MonoAndroid70;MonoAndroid71;MonoAndroid80;MonoAndroid81;MonoAndroid90;</TargetFrameworks>
+    <TargetFrameworks></TargetFrameworks>
+    <TargetFrameworks Condition=" '$(OS)' == 'Windows_NT' ">netstandard1.0;netstandard2.0;Xamarin.iOS10;MonoAndroid60;MonoAndroid70;MonoAndroid71;MonoAndroid80;MonoAndroid81;MonoAndroid90;uap10.0.16299;tizen40;</TargetFrameworks>
```
Comment (Redth) [inline Xamarin.Essentials/Xamarin.Essentials.csproj]: @jamesmontemagno I like that this is clean, but by doing this we require everyone who builds essentials to have Tizen SDK installed.  Do we want to do this? or should we add a separate project (and therefore separate nupkg :/ )

Comment (rookiejava) [reply Xamarin.Essentials/Xamarin.Essentials.csproj]: Do not need to install Tizen SDK to build tizen project.

Tizen TFM is already integrated into the Nuget.Client.

 Tizen SDK installation is only required if you want to use the emulator or device manager, and so on.

File: Xamarin.Essentials/Barometer/Barometer.tizen.cs (hunk 2 above)
Comment (rookiejava) [reply Xamarin.Essentials/Barometer/Barometer.tizen.cs]: Yes, It is `hPa (hectopascals)`. 
https://developer.tizen.org/ko/development/guides/.net-application/location-and-sensors/device-sensors#pressure

File: Xamarin.Essentials/Geolocation/Geolocation.tizen.cs (hunk 3 above)
Comment (rookiejava) [reply Xamarin.Essentials/Geolocation/Geolocation.tizen.cs]: - Accuracy : meters
- Altitude : meters
- Speed : km/h

https://samsung.github.io/TizenFX/API4/api/Tizen.Location.Location.html

File: Xamarin.Essentials/Magnetometer/Magnetometer.tizen.cs (hunk 4 above)
Comment (rookiejava) [reply Xamarin.Essentials/Magnetometer/Magnetometer.tizen.cs]: Yes. It is `µT (microteslas)`

https://developer.tizen.org/ko/development/guides/.net-application/location-and-sensors/device-sensors#magnetic

File: Xamarin.Essentials/TextToSpeech/TextToSpeech.tizen.cs (hunk 1 above)
Comment (rookiejava) [reply Xamarin.Essentials/TextToSpeech/TextToSpeech.tizen.cs]: the default value of speed is `0` (it means `auto`) in tizen. So, we'd like to set speed only if pitch value is specified by user. (like android implementation as below)

https://github.com/xamarin/Essentials/blob/05ba0792df416f38df998c26488ba92c3f5c90ab/Xamarin.Essentials/TextToSpeech/TextToSpeech.android.cs#L143-L147

File: Xamarin.Essentials/Geolocation/Geolocation.tizen.cs (hunk 3 above)
Comment (Redth) [reply Xamarin.Essentials/Geolocation/Geolocation.tizen.cs]: Can you please translate the value to meters / second to match the other platforms?

File: Xamarin.Essentials/TextToSpeech/TextToSpeech.tizen.cs (hunk 1 above)
Comment (Redth) [reply Xamarin.Essentials/TextToSpeech/TextToSpeech.tizen.cs]: It's important that the values that will be set in the cross platform API for essentials translate into the correct _relative_ values in the native SDK.  Here's what I'm thinking might work for Tizen in this case:

```csharp
var TIZEN_MAX = 100;
var TIZEN_MIN = -100;
var TIZEN_DEFAULT = 0;

var ESSENTIALS_MAX = 2.0f;
var ESSENTIALS_MIN = 0.0f;
var ESSENTIALS_DEFAULT = 1.0f;

var tizenPitch = 0;

if (essentialsPitch < 0f)
    essentialsPitch = 0f;
if (essentialsPitch > 2.0f)
    essentialsPitch = 2.0f;

if (essentialsPitch < ESSENTIALS_DEFAULT)
    tizenPitch = (ESSENTIALS_MIN + essentialsPitch) * TIZEN_MIN;
else if (essentialsPitch > ESSENTIALS_DEFAULT)
    tizenPitch = (ESSENTIALS_MAX - essentialsPitch) * TIZEN_MAX;
```

I'm not sure what the real min/max are for Tizen's values but that code should roughly do the right thing, where in the end you'll want to set `tts.SetPitch(tizenPitch)`

File: Xamarin.Essentials/Geolocation/Geolocation.tizen.cs (hunk 3 above)
Comment (Redth) [reply Xamarin.Essentials/Geolocation/Geolocation.tizen.cs]: I think this is the correct calculation to use, to set your `Location.Speed`

```csharp
double KmToMetersPerSecond(double km)
    => km * 0.277778;
```

File: Xamarin.Essentials/TextToSpeech/TextToSpeech.tizen.cs (hunk 1 above)
Comment (rookiejava) [reply Xamarin.Essentials/TextToSpeech/TextToSpeech.tizen.cs]: Agreed. (`min` is `1`, `max` is `15` for pitch on tizen for TTS. ).

File: Xamarin.Essentials/Geolocation/Geolocation.tizen.cs (hunk 3 above)
Comment (rookiejava) [reply Xamarin.Essentials/Geolocation/Geolocation.tizen.cs]: You are right. We will check the actual values once more (to verify whether the document is in error or not) and update them immediately.

File: Xamarin.Essentials/TextToSpeech/TextToSpeech.tizen.cs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+                    {
+                        language = voice.Language;
+                        voiceType = voice.VoiceType;
+                    }
+                }
+            }
+
+            var pitch = tts.GetSpeedRange().Normal;
+            if (options?.Pitch.HasValue ?? false)
+                pitch = (int)(options.Pitch.Value / PitchMax * tts.GetSpeedRange().Max);
```
Comment (rookiejava) [inline Xamarin.Essentials/TextToSpeech/TextToSpeech.tizen.cs]: Rounding convention is required to this value.

File: Xamarin.Essentials/TextToSpeech/TextToSpeech.tizen.cs [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+                {
+                    if (voice.Language == options.Locale.Language)
+                    {
+                        language = voice.Language;
+                        voiceType = voice.VoiceType;
+                    }
+                }
+            }
+
+            var pitch = tts.GetSpeedRange().Normal;
```
Comment (rookiejava) [inline Xamarin.Essentials/TextToSpeech/TextToSpeech.tizen.cs]: Default value = 0 (AUTO) is OK.

Comment (sung-su) [reply Xamarin.Essentials/TextToSpeech/TextToSpeech.tizen.cs]: done.

File: Xamarin.Essentials/TextToSpeech/TextToSpeech.tizen.cs (hunk 6 above)
Comment (sung-su) [reply Xamarin.Essentials/TextToSpeech/TextToSpeech.tizen.cs]: done

**Pull request merged**